google/cloud/forseti/services/scanner/dao.py

```diff
@@ -190,3 +202,39 @@ def map_by_resource(violation_rows):
             v_by_type[v_resource].append(v_data)
 
     return dict(v_by_type)
+
+def _create_violation_hash(violation_full_name, inventory_data, violation_data,
+                           algorithm=FALLBACK_HASH_ALGORITHM):
+    """Create a hash of violation data using the requested algorithm.
+
+    Args:
+        violation_full_name (str): The full name of the violation.
+        inventory_data (str): The inventory data.
+        violation_data (dict): A violation.
+        algorithm (str): The algorithm used to create the hash.
+
+    Returns:
+        str: The resulting hex digest or '' if we can't successfully create
+        a hash.
+    """
+    try:
+        violation_hash = hashlib.new(algorithm)
+    except ValueError as e:
+        LOGGER.error('Unable to use the specified hash algorithm: %s',
+                     algorithm)
+        LOGGER.info('Using the fallback hash algorithm: %s',
+                    FALLBACK_HASH_ALGORITHM)
+        violation_hash = hashlib.new(FALLBACK_HASH_ALGORITHM)
+
+    try:
+        violation_hash.update(
+            violation_full_name +
+            inventory_data +
+            json.dumps(violation_data)
+        )
+    except TypeError as e:
```

Comment:
You will also get a NoneType exception if any of the passed in args are None, which they will be if you use .get() in create() above without setting the default value to an empty string.

Reply:
Nevermind, you will always get a TypeError here. The rest of the comment is still valid.

Reply:
Nice catch, fixed.